onnxruntime/core/platform/windows/env.cc

```diff
@@ -148,28 +171,141 @@ class WindowsEnv : public Env {
+    KAFFINITY group_id = 0;
+    for (const auto& group_info : this->group_info_vec_) {
+      size_t bit_offset = 0;
+      constexpr size_t bit_limit = sizeof(KAFFINITY) << 3;
+      KAFFINITY active_processor_mask = group_info.ActiveProcessorMask;
+      for (int i = 0; i < (group_info.ActiveProcessorCount >> 1); ++i) {
+        KAFFINITY processor_mask = 0;
+        size_t processor_count = 0;
+        while (processor_count < 2 && bit_offset < bit_limit) {
+          auto current_bit = (0x1 << bit_offset);
+          if (active_processor_mask & current_bit) {
+            processor_mask |= current_bit;
+            processor_count++;
+          }
+          bit_offset++;
+        }
+        kAffinities.push_back(group_id);
+        kAffinities.push_back(processor_mask);
+      }
+      group_id++;
     }
-    std::vector<size_t> ret;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
-      if (buffer[i].Relationship == RelationProcessorCore) {
-        ret.push_back(buffer[i].ProcessorMask);
+    if (kAffinities.empty()) {
+      return std::thread::hardware_concurrency() >> 1;
+    }
+    size_t num_of_size_t = kAffinities.size() * sizeof(KAFFINITY) / sizeof(size_t);
+    affinities.clear();
+    affinities.resize(num_of_size_t);
+    memcpy(affinities.data(), kAffinities.data(), num_of_size_t * sizeof(size_t));
+    return kAffinities.size();
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(int procssor_from, int processor_to) {
+    if (procssor_from > processor_to) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> must be smaller or equal to <to>";
+      return {-1, 0};
+    }
+    int processor_id = 1;
+    int processor_count = 0;
+    KAFFINITY group_id = 0;
+    KAFFINITY processor_mask = 0;
+    for (const auto& group_info : this->group_info_vec_) {
+      for (int i = 0; i < group_info.ActiveProcessorCount; ++i, ++processor_id) {
+        if (processor_id >= procssor_from && processor_id <= processor_to) {
+          processor_mask |= 1ULL << i;
+          processor_count += 1;
+        }
+      }
+      if (processor_mask) {
+        break;
       }
+      group_id++;
     }
-    if (ret.empty())
-      return generate_vector_of_n(std::thread::hardware_concurrency());
-    return ret;
+    if (processor_count < processor_to - procssor_from + 1) {
+      LOGS_DEFAULT(ERROR) << "Processor <from> or <to> crossed group boundary";
+      return {-1,0};
+    }
+    return {group_id, processor_mask};
+  }
+
+  std::pair<KAFFINITY, KAFFINITY> GetGroupAffinity(const std::vector<std::string>& processor_id_strs) {
+    if (processor_id_strs.empty()) {
+      return {-1, 0};
+    }
+    std::set<int> processor_ids;
+    std::for_each(processor_id_strs.begin(),
+                  processor_id_strs.end(), [&](const std::string& processor_id_str) {
+                    processor_ids.insert( std::stoi(processor_id_str.c_str()));
```

Comment:
what if stoi throws?

Reply:
At the end of function "ReadThreadAffinityConfig" we are capturing the exceptions coming out of stoi.